src/abstract-interpretation/processor.ts

```diff
@@ -1,7 +1,157 @@
-import { DataflowInformation } from '../dataflow/internal/info'
-import { NormalizedAst } from '../r-bridge'
+import {DataflowInformation} from '../dataflow/internal/info'
+import {NodeId, NormalizedAst, ParentInformation, RBinaryOp, RType} from '../r-bridge'
+import {CfgVertexType, extractCFG} from '../util/cfg/cfg'
+import {visitCfg} from '../util/cfg/visitor'
+import {guard} from '../util/assert'
+import {DataflowGraphVertexInfo, EdgeType, OutgoingEdges} from '../dataflow'
+
+const constraintMap = new Map<NodeId, Constraints>()
+
+class Stack<ElementType> {
+	private backingStore: ElementType[] = []
+
+	size(): number { return this.backingStore.length }
+	peek(): ElementType | undefined { return this.backingStore[this.size() - 1] }
+	pop(): ElementType | undefined { return this.backingStore.pop() }
+	push(item: ElementType): ElementType {
+		this.backingStore.push(item)
+		return item
+	}
+}
+
+interface IHandler<ValueType> {
+	name:  string,
+	enter: () => void
+	exit:  () => ValueType
+	next:  (value: ValueType) => void
+}
+
+type Intervals = Set<{min: number, minInclusive: boolean, max: number, maxInclusive: boolean}>
+
+interface Constraints {
+	node:      NodeId,
+	intervals: Intervals,
+	debugMsg:  string
+}
+
+function intervalFromNumber(n: number): Intervals {
+	return new Set([{min: n, minInclusive: true, max: n, maxInclusive: true}])
+}
+
+function getIntervalsOfDfgChild(node: NodeId, dfg: DataflowInformation): Intervals {
+	const dfgNode: [DataflowGraphVertexInfo, OutgoingEdges] | undefined = dfg.graph.get(node)
+	guard(dfgNode !== undefined, `No DFG-Node found with ID ${node}`)
+	const [_, children] = dfgNode
+	const ids = Array.from(children.entries())
+		.filter(([_, edge]) => edge.types.has(EdgeType.Reads))
+		.map(([id, _]) => id)
+	if(ids.length === 0) {
+		return new Set()
+	}
+	guard(ids.length === 1, `Multiple reads-edges found for node ${node}`)
+	const id = ids[0]
+	const constraint = constraintMap.get(id)
+	guard(constraint !== undefined, `No constraint found for ID ${id}`)
+	return constraint.intervals
+}
+
+class Assignment implements IHandler<Constraints> {
+	private lhs:           NodeId | undefined
+	private rhs:           NodeId | undefined
+	private readonly node: RBinaryOp<ParentInformation>
+	name = 'Assignment'
+
+	constructor(node: RBinaryOp<ParentInformation>) {
+		this.node = node
+	}
+
+	enter(): void {
+		console.log(`Entered ${this.name} ${this.node.info.id}`)
+	}
+
+	exit(): Constraints {
+		console.log(`Exited ${this.name} ${this.node.info.id}`)
+		return {
+			node:      this.lhs as NodeId,
+			intervals: new Set(), // TODO: check interval of the rhs
```

Comment:
in general, not just the rhs but the _source_ of the assignment (`x <- source`, `source -> x`). This will be easier with the dataflow v2 update which will just have one ""assign"" it handles.

Reply:
Oh yeah, I forgot that the assignments source must not always be on the rhs in R.